src/optimizer/optimizer.cpp

```diff
@@ -87,117 +87,125 @@ unique_ptr<LogicalOperator> Optimizer::Optimize(unique_ptr<LogicalOperator> plan
+		RunOptimizer(OptimizerType::JOIN_ORDER, [&]() {
+			JoinOrderOptimizer optimizer(context);
+			plan = optimizer.Optimize(std::move(plan));
+		});
+
+		// rewrites UNNESTs in DelimJoins by moving them to the projection
+		RunOptimizer(OptimizerType::UNNEST_REWRITER, [&]() {
+			UnnestRewriter unnest_rewriter;
+			plan = unnest_rewriter.Optimize(std::move(plan));
+		});
+
+		// removes unused columns
+		RunOptimizer(OptimizerType::UNUSED_COLUMNS, [&]() {
+			RemoveUnusedColumns unused(binder, context, true);
+			unused.VisitOperator(*plan);
+		});
+
+		// Remove duplicate groups from aggregates
+		RunOptimizer(OptimizerType::DUPLICATE_GROUPS, [&]() {
+			RemoveDuplicateGroups remove;
+			remove.VisitOperator(*plan);
+		});
+
+		// then we extract common subexpressions inside the different operators
+		RunOptimizer(OptimizerType::COMMON_SUBEXPRESSIONS, [&]() {
+			CommonSubExpressionOptimizer cse_optimizer(binder);
+			cse_optimizer.VisitOperator(*plan);
+		});
+
+		// creates projection maps so unused columns are projected out early
+		RunOptimizer(OptimizerType::COLUMN_LIFETIME, [&]() {
+			ColumnLifetimeAnalyzer column_lifetime(true);
+			column_lifetime.VisitOperator(*plan);
+		});
+
+		// perform statistics propagation
+		column_binding_map_t<unique_ptr<BaseStatistics>> statistics_map;
+		RunOptimizer(OptimizerType::STATISTICS_PROPAGATION, [&]() {
+			StatisticsPropagator propagator(*this);
+			propagator.PropagateStatistics(plan);
+			statistics_map = propagator.GetStatisticsMap();
+		});
+
+		// remove duplicate aggregates
+		RunOptimizer(OptimizerType::COMMON_AGGREGATE, [&]() {
+			CommonAggregateOptimizer common_aggregate;
+			common_aggregate.VisitOperator(*plan);
 		});
+
+		// creates projection maps so unused columns are projected out early
+		RunOptimizer(OptimizerType::COLUMN_LIFETIME, [&]() {
+			ColumnLifetimeAnalyzer column_lifetime(true);
+			column_lifetime.VisitOperator(*plan);
+		});
+
+		// compress data based on statistics for materializing operators
+		RunOptimizer(OptimizerType::COMPRESSED_MATERIALIZATION, [&]() {
+			CompressedMaterialization compressed_materialization(context, binder, std::move(statistics_map));
+			compressed_materialization.Compress(plan);
+		});
+
+		// transform ORDER BY + LIMIT to TopN
+		RunOptimizer(OptimizerType::TOP_N, [&]() {
+			TopN topn;
+			plan = topn.Optimize(std::move(plan));
+		});
+
+		// apply simple expression heuristics to get an initial reordering
+		RunOptimizer(OptimizerType::REORDER_FILTER, [&]() {
+			ExpressionHeuristics expression_heuristics(*this);
+			plan = expression_heuristics.Rewrite(std::move(plan));
+		});
+
+		for (auto &optimizer_extension : DBConfig::GetConfig(context).optimizer_extensions) {
+			RunOptimizer(OptimizerType::EXTENSION, [&]() {
+				// if any of the optimizer extensions requested a rerun of the optimizer, we set the flag here
+				run_optimizer = run_optimizer || optimizer_extension.optimize_function(
```

Comment:
I am not sure I see what's the termination condition here, say an extensions does minor modifications and returns true, are we going to keep going?
And the plan will keep improving, get to a fixed point, or to some weird state that we never accounted for.

Proper solution might require some more infrastructure (I spent some time on LLVM pass manager, it's a fun topic) (also it's not a given that it makes sense to run optimizers in a loop), but maybe for now saying that the whole block is executed at maximum 2 or 3 times might do it?

Reply:
So it's not necessary to always rerun if you do a modification (that's how it currently works), the idea with this is just that you can opt into rerunning if you end up doing a massive change to the plan (e.g. adding completely new operators, making the plan "larger" = less optimized). 

I don't think its gonna be very common for extension optimizers to rerun more than 1 or 2 times at most.
In the future if we end up with a lot of extension optimizers we may want to look into doing something like the pass manager where an optimizer can request to schedule additional specific passes, but I think for now this is fine.

I can add an upper bound for safety (e.g. 99). This might also be something we could allow controlling through a config option in the future. (e.g. MAX_ADDITIONAL_OPTIMIZER_PASSES)

Reply:
Upper bound also good, main worry is having a way to make it terminate even in the case the extension optimizer is implemented as `{return true;}`